Source/Core/VideoCommon/LightingShaderGen.cpp

```diff
@@ -83,94 +104,141 @@ static void GenerateLightShader(ShaderCode& object, const LightingUidData& uid_d
 {
-  for (u32 j = 0; j < NUM_XF_COLOR_CHANNELS; j++)
+  for (u32 chan = 0; chan < NUM_XF_COLOR_CHANNELS; chan++)
   {
-    uid_data.matsource |= static_cast<u32>(xfmem.color[j].matsource.Value()) << j;
-    uid_data.matsource |= static_cast<u32>(xfmem.alpha[j].matsource.Value()) << (j + 2);
-    uid_data.enablelighting |= xfmem.color[j].enablelighting << j;
-    uid_data.enablelighting |= xfmem.alpha[j].enablelighting << (j + 2);
+    // Data for alpha is stored after all colors
+    const u32 chan_a = chan + NUM_XF_COLOR_CHANNELS;
+
+    uid_data.matsource |= static_cast<u32>(xfmem.color[chan].matsource.Value()) << chan;
+    uid_data.matsource |= static_cast<u32>(xfmem.alpha[chan].matsource.Value()) << chan_a;
+    uid_data.enablelighting |= xfmem.color[chan].enablelighting << chan;
+    uid_data.enablelighting |= xfmem.alpha[chan].enablelighting << chan_a;
 
-    if ((uid_data.enablelighting & (1 << j)) != 0)  // Color lights
+    if ((uid_data.enablelighting & (1 << chan)) != 0)  // Color lights
     {
-      uid_data.ambsource |= static_cast<u32>(xfmem.color[j].ambsource.Value()) << j;
-      uid_data.attnfunc |= static_cast<u32>(xfmem.color[j].attnfunc.Value()) << (2 * j);
-      uid_data.diffusefunc |= static_cast<u32>(xfmem.color[j].diffusefunc.Value()) << (2 * j);
-      uid_data.light_mask |= xfmem.color[j].GetFullLightMask() << (8 * j);
+      uid_data.ambsource |= static_cast<u32>(xfmem.color[chan].ambsource.Value()) << chan;
+      const u32 attn = (static_cast<u32>(xfmem.color[chan].attnfunc.Value()) << 1) |
+                       static_cast<u32>(xfmem.color[chan].attnenable.Value());
+      uid_data.attnfunc |= attn << (2 * chan);
+      uid_data.diffusefunc |= static_cast<u32>(xfmem.color[chan].diffusefunc.Value()) << (2 * chan);
+      uid_data.light_mask |= xfmem.color[chan].GetFullLightMask() << (NUM_XF_LIGHTS * chan);
     }
-    if ((uid_data.enablelighting & (1 << (j + 2))) != 0)  // Alpha lights
+    if ((uid_data.enablelighting & (1 << chan_a)) != 0)  // Alpha lights
     {
-      uid_data.ambsource |= static_cast<u32>(xfmem.alpha[j].ambsource.Value()) << (j + 2);
-      uid_data.attnfunc |= static_cast<u32>(xfmem.alpha[j].attnfunc.Value()) << (2 * (j + 2));
-      uid_data.diffusefunc |= static_cast<u32>(xfmem.alpha[j].diffusefunc.Value()) << (2 * (j + 2));
-      uid_data.light_mask |= xfmem.alpha[j].GetFullLightMask() << (8 * (j + 2));
+      uid_data.ambsource |= static_cast<u32>(xfmem.alpha[chan].ambsource.Value()) << chan_a;
+      const u32 attn = (static_cast<u32>(xfmem.alpha[chan].attnfunc.Value()) << 1) |
+                       static_cast<u32>(xfmem.alpha[chan].attnenable.Value());
+      uid_data.attnfunc |= attn << (2 * chan_a);
+      uid_data.diffusefunc |= static_cast<u32>(xfmem.alpha[chan].diffusefunc.Value())
+                              << (2 * chan_a);
+      uid_data.light_mask |= xfmem.alpha[chan].GetFullLightMask() << (NUM_XF_LIGHTS * chan_a);
     }
   }
```

Comment:
I don't like the fiddlyness of this function and the lack of type safety (the earlier fifoci differences were from me messing up `attnfunc`.  I would much rather have something like this:

```C++
void GetLightingShaderUid(LightingUidData& uid_data)
{
  for (u32 chan = 0; chan < NUM_XF_COLOR_CHANNELS; chan++)
  {
    uid_data.matsource_color[chan] = xfmem.color[chan].matsource;
    uid_data.matsource_alpha[chan] = xfmem.alpha[chan].matsource;
    uid_data.enablelighting_color[chan] = xfmem.color[chan].enablelighting;
    uid_data.enablelighting_alpha[chan] = xfmem.alpha[chan].enablelighting;

    if (uid_data.enablelighting_color[chan])  // Color lights
    {
      uid_data.ambsource_color[chan] = xfmem.color[chan].ambsource;
      uid_data.attnfunc_color[chan] = xfmem.color[chan].attnfunc;
      uid_data.diffusefunc_color[chan] = xfmem.color[chan].diffusefunc;
      uid_data.light_mask_color[chan] = xfmem.color[chan].GetFullLightMask();
    }
    if (uid_data.enablelighting_alpha[chan])
    {
      uid_data.ambsource_alpha[chan] = xfmem.alpha[chan].ambsource;
      uid_data.attnfunc_alpha[chan] = xfmem.alpha[chan].attnfunc;
      uid_data.diffusefunc_alpha[chan] = xfmem.alpha[chan].diffusefunc;
      uid_data.light_mask_alpha[chan] = xfmem.alpha[chan].GetFullLightMask();
    }
  }
}
```

corresponding to this:

```C++
struct LightingUidData
{
  static_assert(NUM_XF_COLOR_CHANNELS == 2);  // Bit count choices will break otherwise
  static_assert(NUM_XF_LIGHTS == 8);
  union
  {
    BitFieldArray<0, 1, NUM_XF_COLOR_CHANNELS, MatSource, u32> matsource_color;        // 2x1 bit
    BitFieldArray<2, 1, NUM_XF_COLOR_CHANNELS, MatSource, u32> matsource_alpha;        // 2x1 bit
    BitFieldArray<4, 1, NUM_XF_COLOR_CHANNELS, bool, u32> enablelighting_color;        // 2x1 bit
    BitFieldArray<6, 1, NUM_XF_COLOR_CHANNELS, bool, u32> enablelighting_alpha;        // 2x1 bit
    BitFieldArray<8, 1, NUM_XF_COLOR_CHANNELS, AmbSource, u32> ambsource_color;        // 2x1 bit
    BitFieldArray<10, 1, NUM_XF_COLOR_CHANNELS, AmbSource, u32> ambsource_alpha;       // 2x1 bit
    BitFieldArray<12, 2, NUM_XF_COLOR_CHANNELS, DiffuseFunc, u32> diffusefunc_color;   // 2x2 bit
    BitFieldArray<16, 2, NUM_XF_COLOR_CHANNELS, DiffuseFunc, u32> diffusefunc_alpha;   // 2x2 bit
    BitFieldArray<20, 2, NUM_XF_COLOR_CHANNELS, AttenuationFunc, u32> attnfunc_color;  // 2x2 bit
    BitFieldArray<24, 2, NUM_XF_COLOR_CHANNELS, AttenuationFunc, u32> attnfunc_alpha;  // 2x2 bit
  };
  union
  {
    BitFieldArray<0, NUM_XF_LIGHTS, NUM_XF_COLOR_CHANNELS, u32> light_mask_color;
    BitFieldArray<16, NUM_XF_LIGHTS, NUM_XF_COLOR_CHANNELS, u32> light_mask_alpha;
  };
};
```

(This is an earlier draft before I split `AttenuationFunc`.  I'm not sure how I'd deal with the split version, as I am trying to maintain compatibility with existing UIDs).

The problem with doing this is that `BitField` and `BitFieldArray` have deleted copy-assignment operators, as the default copy-assignment operator would copy all bits, and a non-trivial copy-assignment operator messes with unions.  However, even it still is impossible to perform copy-assignment, which messes up the shader UID system (`BitField` can't be used with it at all).  I'm not sure what to do about it.